From 32fa1681d4582c89429b4d58a65492cd1e92ea38 Mon Sep 17 00:00:00 2001
From: "xiaoya.lin" <xiaoya.lin@amlogic.com>
Date: Fri, 7 Jul 2023 18:22:32 +0800
Subject: [PATCH 07/10] encoderhal: Add libamvenc, gstamlvenc plugin [1/1]

PD#SWPL-132055

Problem:
Add encoderhal

Solution:
Add libamvenc, gstamlvenc plugin

Verify:
Local

Signed-off-by: xiaoya.lin <xiaoya.lin@amlogic.com>
Change-Id: I57022ee631c2252d59d7a67afbaad3c43d5b71f3
---
 conf/machine/include/amlogic_externalsrc.inc  |  4 +++
 conf/machine/mesong12b-w400-debian-5.15.conf  |  5 +--
 .../machine/mesong12b-w400-weston-5.15.config |  5 +--
 conf/machine/mesonsc2-ah212-weston.config     |  5 +--
 conf/machine/mesonsm1-ac200-debian-5.15.conf  |  5 +--
 .../machine/mesonsm1-ac200-weston-5.15.config |  5 +--
 conf/machine/mesont7-an400-debian-5.15.conf   |  1 +
 conf/machine/mesont7-an400-debian.config      |  1 +
 conf/machine/mesont7-an400-weston-5.15.config |  1 +
 conf/machine/mesont7-an400-weston.config      |  1 +
 .../machine/mesont7c-an400-weston-5.15.config |  1 +
 conf/machine/mesont7c-an400-weston.config     |  1 +
 .../images/packagegroup-amlogic-baserootfs.bb |  7 ++--
 recipes-kernel/libamvenc-264/h264bitstream.bb | 14 ++++++++
 recipes-kernel/libamvenc-264/libamvenc-264.bb | 31 +++++++++++++++++
 recipes-kernel/libamvenc-265/libamvenc-265.bb | 32 ++++++++++++++++++
 recipes-kernel/libamvenc/libamvenc.bb         | 31 +++++++++++++++++
 .../gst-plugin-venc-amlvenc.bb                | 33 +++++++++++++++++++
 18 files changed, 170 insertions(+), 13 deletions(-)
 create mode 100755 recipes-kernel/libamvenc-264/h264bitstream.bb
 create mode 100755 recipes-kernel/libamvenc-264/libamvenc-264.bb
 create mode 100755 recipes-kernel/libamvenc-265/libamvenc-265.bb
 create mode 100755 recipes-kernel/libamvenc/libamvenc.bb
 create mode 100755 recipes-multimedia/gst-plugin-venc/gst-plugin-venc-amlvenc.bb

diff --git a/conf/machine/include/amlogic_externalsrc.inc b/conf/machine/include/amlogic_externalsrc.inc
index 81ba239b..7db7e63b 100644
--- a/conf/machine/include/amlogic_externalsrc.inc
+++ b/conf/machine/include/amlogic_externalsrc.inc
@@ -123,6 +123,9 @@ EXTERNALSRC_pn-libgdc = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogi
 EXTERNALSRC_pn-libvpcodec = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/h264' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/h264')) else ''}"
 EXTERNALSRC_pn-libvphevcodec = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/h265' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/h265')) else ''}"
 EXTERNALSRC_pn-libjpeg = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/jpeg' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/jpeg')) else ''}"
+EXTERNALSRC_pn-libamvenc = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc')) else ''}"
+EXTERNALSRC_pn-libamvenc-264 = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc_264' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc_264')) else ''}"
+EXTERNALSRC_pn-libamvenc-265 = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc_265' if(os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/libencoder/amvenc_265')) else ''}"
 EXTERNALSRC_pn-mfrlib-amlogic = "${@['', '${MESON_ROOT_PATH}/aml-comp/thirdparty/Mfrlib'][os.path.isdir('${MESON_ROOT_PATH}/aml-comp/thirdparty/Mfrlib')]}"
 EXTERNALSRC_pn-avs-sdk= "${@'${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/avs-sdk' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/avs-sdk')) else ''}"
 EXTERNALSRC_pn-ledring= "${@'${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/ledring' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/ledring')) else ''}"
@@ -154,6 +157,7 @@ EXTERNALSRC_pn-gst-plugin-venc-jpeg = "${@'${MESON_ROOT_PATH}/aml-comp/multimedi
 EXTERNALSRC_pn-gst-plugin-venc-h264 = "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/h264-hcodec' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/h264-hcodec')) else ''}"
 EXTERNALSRC_pn-gst-plugin-venc-h265 = "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/h265-wave420' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/h265-wave420')) else ''}"
 EXTERNALSRC_pn-gst-plugin-venc-multienc = "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/multienc-wave521' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/multienc-wave521')) else ''}"
+EXTERNALSRC_pn-gst-plugin-venc-amlvenc = "${@'${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/amlvenc' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/multimedia/gst-plugin-venc/amlvenc')) else ''}"
 EXTERNALSRC_pn-adla = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/adla' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/adla')) else ''}"
 EXTERNALSRC_pn-npu-common = "${@'${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/aml_npu_common' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/vendor/amlogic/aml_npu_common')) else ''}"
 EXTERNALSRC_pn-npu = "${@'${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/npu/nanoq' if (os.path.isdir('${MESON_ROOT_PATH}/aml-comp/hardware/aml-5.4/amlogic/npu/nanoq')) else ''}"
diff --git a/conf/machine/mesong12b-w400-debian-5.15.conf b/conf/machine/mesong12b-w400-debian-5.15.conf
index f7453e37..8770b2ce 100644
--- a/conf/machine/mesong12b-w400-debian-5.15.conf
+++ b/conf/machine/mesong12b-w400-debian-5.15.conf
@@ -62,8 +62,9 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
-DISTRO_FEATURES_append = " aml-libvpcodec"
-DISTRO_FEATURES_append = " aml-libvphevcodec"
+DISTRO_FEATURES_append = " aml-libamvenc-264"
+DISTRO_FEATURES_append = " aml-libamvenc-265"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " arm-isp "
 DISTRO_FEATURES_append = " softap"
diff --git a/conf/machine/mesong12b-w400-weston-5.15.config b/conf/machine/mesong12b-w400-weston-5.15.config
index 840b9b65..4aacf762 100644
--- a/conf/machine/mesong12b-w400-weston-5.15.config
+++ b/conf/machine/mesong12b-w400-weston-5.15.config
@@ -62,8 +62,9 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
-DISTRO_FEATURES_append = " aml-libvpcodec"
-DISTRO_FEATURES_append = " aml-libvphevcodec"
+DISTRO_FEATURES_append = " aml-libamvenc-264"
+DISTRO_FEATURES_append = " aml-libamvenc-265"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " arm-isp "
 DISTRO_FEATURES_append = " softap"
diff --git a/conf/machine/mesonsc2-ah212-weston.config b/conf/machine/mesonsc2-ah212-weston.config
index 4fcb7b57..91da8cdc 100644
--- a/conf/machine/mesonsc2-ah212-weston.config
+++ b/conf/machine/mesonsc2-ah212-weston.config
@@ -61,8 +61,9 @@ DISTRO_FEATURES_append = " aml-dtv "
 #DISTRO_FEATURES_append = " recovery "
 #DISTRO_FEATURES_append = " swupdate "
 DISTRO_FEATURES_append = " aml-libjpeg "
-DISTRO_FEATURES_append = " aml-libvpcodec "
-DISTRO_FEATURES_append = " aml-libvphevcodec "
+DISTRO_FEATURES_append = " aml-libamvenc-264"
+DISTRO_FEATURES_append = " aml-libamvenc-265"
+DISTRO_FEATURES_append = " aml-libamvenc"
 DISTRO_FEATURES_append = " gst-plugin-venc "
 
 OVERRIDES .= ":ah212:sc2:onepass"
diff --git a/conf/machine/mesonsm1-ac200-debian-5.15.conf b/conf/machine/mesonsm1-ac200-debian-5.15.conf
index 9472cfa9..30ab2e10 100644
--- a/conf/machine/mesonsm1-ac200-debian-5.15.conf
+++ b/conf/machine/mesonsm1-ac200-debian-5.15.conf
@@ -61,8 +61,9 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
-DISTRO_FEATURES_append = " aml-libvpcodec"
-DISTRO_FEATURES_append = " aml-libvphevcodec"
+DISTRO_FEATURES_append = " aml-libamvenc-264"
+DISTRO_FEATURES_append = " aml-libamvenc-265"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " softap"
 DISTRO_FEATURES_append = " aml-libgdc"
diff --git a/conf/machine/mesonsm1-ac200-weston-5.15.config b/conf/machine/mesonsm1-ac200-weston-5.15.config
index 6b15b887..8e1000d4 100644
--- a/conf/machine/mesonsm1-ac200-weston-5.15.config
+++ b/conf/machine/mesonsm1-ac200-weston-5.15.config
@@ -61,8 +61,9 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
-DISTRO_FEATURES_append = " aml-libvpcodec"
-DISTRO_FEATURES_append = " aml-libvphevcodec"
+DISTRO_FEATURES_append = " aml-libamvenc-264"
+DISTRO_FEATURES_append = " aml-libamvenc-265"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " softap"
 DISTRO_FEATURES_append = " aml-libgdc"
diff --git a/conf/machine/mesont7-an400-debian-5.15.conf b/conf/machine/mesont7-an400-debian-5.15.conf
index 9289d4df..66d9941e 100644
--- a/conf/machine/mesont7-an400-debian-5.15.conf
+++ b/conf/machine/mesont7-an400-debian-5.15.conf
@@ -66,6 +66,7 @@ DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/conf/machine/mesont7-an400-debian.config b/conf/machine/mesont7-an400-debian.config
index b75f6297..80ab401a 100644
--- a/conf/machine/mesont7-an400-debian.config
+++ b/conf/machine/mesont7-an400-debian.config
@@ -60,6 +60,7 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/conf/machine/mesont7-an400-weston-5.15.config b/conf/machine/mesont7-an400-weston-5.15.config
index 7ed2ba58..f18627b0 100644
--- a/conf/machine/mesont7-an400-weston-5.15.config
+++ b/conf/machine/mesont7-an400-weston-5.15.config
@@ -66,6 +66,7 @@ DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/conf/machine/mesont7-an400-weston.config b/conf/machine/mesont7-an400-weston.config
index 96d311f5..f25c900c 100644
--- a/conf/machine/mesont7-an400-weston.config
+++ b/conf/machine/mesont7-an400-weston.config
@@ -60,6 +60,7 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/conf/machine/mesont7c-an400-weston-5.15.config b/conf/machine/mesont7c-an400-weston-5.15.config
index 240bd990..a35a4a99 100755
--- a/conf/machine/mesont7c-an400-weston-5.15.config
+++ b/conf/machine/mesont7c-an400-weston-5.15.config
@@ -65,6 +65,7 @@ DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " gst-plugin-venc"
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp"
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/conf/machine/mesont7c-an400-weston.config b/conf/machine/mesont7c-an400-weston.config
index c04c79e5..9776917b 100644
--- a/conf/machine/mesont7c-an400-weston.config
+++ b/conf/machine/mesont7c-an400-weston.config
@@ -60,6 +60,7 @@ DISTRO_FEATURES_append = " gstreamer1"
 DISTRO_FEATURES_append = " aml-dtv "
 DISTRO_FEATURES_append = " aml-libjpeg"
 DISTRO_FEATURES_append = " aml-libmultienc"
+DISTRO_FEATURES_append = " aml-libamvenc"
 #DISTRO_FEATURES_append = " playready "
 DISTRO_FEATURES_append = " aml-isp arm-isp "
 DISTRO_FEATURES_append = " amlogic-tv "
diff --git a/recipes-core/images/packagegroup-amlogic-baserootfs.bb b/recipes-core/images/packagegroup-amlogic-baserootfs.bb
index 07af3021..2ec6fe5e 100644
--- a/recipes-core/images/packagegroup-amlogic-baserootfs.bb
+++ b/recipes-core/images/packagegroup-amlogic-baserootfs.bb
@@ -89,6 +89,9 @@ RDEPENDS_packagegroup-amlogic-baserootfs = "\
     ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libmultienc', 'libmultienc libge2d libion', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libvpcodec', 'libvpcodec h264bitstream libion', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libvphevcodec', 'libvphevcodec libge2d libion', '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libamvenc-264', 'libamvenc-264 h264bitstream libion', '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libamvenc-265', 'libamvenc-265 libge2d libion', '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'aml-libamvenc', 'libamvenc libion', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'mediactrlsrc', 'mediactrlsrc mediactrlsrc-dev', '', d)} \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gstreamer1', \
         'gstreamer1.0-plugins-good \
@@ -154,9 +157,7 @@ RDEPENDS_packagegroup-amlogic-baserootfs = "\
 #VENC related
 RDEPENDS_packagegroup-amlogic-baserootfs += " \
     ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-venc', bb.utils.contains('DISTRO_FEATURES', 'aml-libjpeg', 'gst-plugin-venc-jpeg', '', d), '', d)} \
-    ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-venc', bb.utils.contains('DISTRO_FEATURES', 'aml-libmultienc', 'gst-plugin-venc-multienc', '', d), '', d)} \
-    ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-venc', bb.utils.contains('DISTRO_FEATURES', 'aml-libvpcodec', 'gst-plugin-venc-h264', '', d), '', d)} \
-    ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-venc', bb.utils.contains('DISTRO_FEATURES', 'aml-libvphevcodec', 'gst-plugin-venc-h265', '', d), '', d)} \
+    ${@bb.utils.contains('DISTRO_FEATURES', 'gst-plugin-venc', bb.utils.contains('DISTRO_FEATURES', 'aml-libamvenc', 'gst-plugin-venc-amlvenc', '', d), '', d)} \
     "
 
 #AML gstreamer aml utils, common lib for other plugins
diff --git a/recipes-kernel/libamvenc-264/h264bitstream.bb b/recipes-kernel/libamvenc-264/h264bitstream.bb
new file mode 100755
index 00000000..3c842fb7
--- /dev/null
+++ b/recipes-kernel/libamvenc-264/h264bitstream.bb
@@ -0,0 +1,14 @@
+SUMMARY = "h264bitstream"
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+inherit autotools pkgconfig
+
+S = "${WORKDIR}/git"
+SRCREV = "34f3c58afa3c47b6cf0a49308a68cbf89c5e0bff"
+SRC_URI = "git://github.com/aizvorski/h264bitstream"
+
+FILES_${PN} = " ${libdir}/* ${bindir}/*"
+FILES_${PN}-dev = "${includedir}/*"
+INSANE_SKIP_${PN} = "dev-so"
+INSANE_SKIP_${PN}-dev = "dev-elf dev-so"
diff --git a/recipes-kernel/libamvenc-264/libamvenc-264.bb b/recipes-kernel/libamvenc-264/libamvenc-264.bb
new file mode 100755
index 00000000..51733d8e
--- /dev/null
+++ b/recipes-kernel/libamvenc-264/libamvenc-264.bb
@@ -0,0 +1,31 @@
+SUMMARY = "aml libamvenc_264 library (H264) for Amlogic HCODEC IP"
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += "h264bitstream libion"
+
+SRCREV ?="${AUTOREV}"
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE += "LIBAMVENC_264_STAGING_DIR=${D} CROSS=${TARGET_PREFIX} TARGET_DIR=${D} STAGING_DIR=${D} DESTDIR=${D} INSTALL_DIR=${D}/usr/lib"
+
+do_compile () {
+    cd ${S}
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/bjunion_enc
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/
+}
+
+do_install () {
+    install -d ${D}${includedir}
+    install -d ${D}${libdir}
+    install -d ${D}${bindir}
+    install -m 0644 -D ${S}/bjunion_enc/vpcodec_1_0.h ${D}${includedir}
+    install -m 0644 -D ${S}/bjunion_enc/libamvenc_264.so ${D}${libdir}
+    install -m 0644 -D ${S}/testApi ${D}${bindir}
+}
+
+FILES_${PN} = " ${libdir}/* ${bindir}/*"
+FILES_${PN}-dev = "${includedir}/*"
+INSANE_SKIP_${PN} = "dev-so"
+INSANE_SKIP_${PN}-dev = "dev-elf dev-so"
diff --git a/recipes-kernel/libamvenc-265/libamvenc-265.bb b/recipes-kernel/libamvenc-265/libamvenc-265.bb
new file mode 100755
index 00000000..9cdbdc7c
--- /dev/null
+++ b/recipes-kernel/libamvenc-265/libamvenc-265.bb
@@ -0,0 +1,32 @@
+SUMMARY = "aml libamvenc_265 library H265 for C&M Wav420 IP"
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += "libion"
+DEPENDS += "libge2d"
+
+SRCREV ?="${AUTOREV}"
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE += "LIBAMVENC_265_STAGING_DIR=${D} CROSS=${TARGET_PREFIX} TARGET_DIR=${D} STAGING_DIR=${D} DESTDIR=${D} INSTALL_DIR=${D}/usr/lib"
+
+do_compile () {
+    cd ${S}
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/EncoderAPI-HEVC/hevc_enc/
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/EncoderAPI-HEVC/
+}
+
+do_install () {
+    install -d ${D}${includedir}
+    install -d ${D}${libdir}
+    install -d ${D}${bindir}
+    install -m 0644 -D ${S}/EncoderAPI-HEVC/hevc_enc/vp_hevc_codec_1_0.h ${D}${includedir}
+    install -m 0644 -D ${S}/EncoderAPI-HEVC/hevc_enc/libamvenc_265.so ${D}${libdir}
+    install -m 0644 -D ${S}/EncoderAPI-HEVC/testHevcApi ${D}${bindir}
+}
+
+FILES_${PN} = " ${libdir}/* ${bindir}/*"
+FILES_${PN}-dev = "${includedir}/*"
+INSANE_SKIP_${PN} = "dev-so"
+INSANE_SKIP_${PN}-dev = "dev-elf dev-so"
diff --git a/recipes-kernel/libamvenc/libamvenc.bb b/recipes-kernel/libamvenc/libamvenc.bb
new file mode 100755
index 00000000..66b2ee5b
--- /dev/null
+++ b/recipes-kernel/libamvenc/libamvenc.bb
@@ -0,0 +1,31 @@
+SUMMARY = "aml libamvenc library (encoder hal) for Amlogic"
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += "libion"
+
+SRCREV ?="${AUTOREV}"
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE += "LIBAMVENC_STAGING_DIR=${D} CROSS=${TARGET_PREFIX} TARGET_DIR=${D} STAGING_DIR=${D} DESTDIR=${D} INSTALL_DIR=${D}/usr/lib"
+
+do_compile () {
+    cd ${S}
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/amvenc_lib
+    oe_runmake ${EXTRA_OEMAKE} -C ${S}/amvenc_test
+}
+
+do_install () {
+    install -d ${D}${includedir}
+    install -d ${D}${libdir}
+    install -d ${D}${bindir}
+    install -m 0644 -D ${S}/amvenc_lib/include/amvenc.h ${D}${includedir}
+    install -m 0644 -D ${S}/amvenc_lib/libamvenc.so ${D}${libdir}
+    install -m 0644 -D ${S}/amvenc_test/amvenc_test ${D}${bindir}
+}
+
+FILES_${PN} = " ${libdir}/* ${bindir}/*"
+FILES_${PN}-dev = "${includedir}/*"
+INSANE_SKIP_${PN} = "dev-so"
+INSANE_SKIP_${PN}-dev = "dev-elf dev-so"
diff --git a/recipes-multimedia/gst-plugin-venc/gst-plugin-venc-amlvenc.bb b/recipes-multimedia/gst-plugin-venc/gst-plugin-venc-amlvenc.bb
new file mode 100755
index 00000000..85278793
--- /dev/null
+++ b/recipes-multimedia/gst-plugin-venc/gst-plugin-venc-amlvenc.bb
@@ -0,0 +1,33 @@
+DESCRIPTION = "Amlogic H264/H265 encode plugin"
+
+LICENSE = "AMLOGIC"
+LIC_FILES_CHKSUM = "file://${COREBASE}/../meta-meson/license/AMLOGIC;md5=6c70138441c57c9e1edb9fde685bd3c8"
+
+DEPENDS += "gstreamer1.0 gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-libav"
+DEPENDS += "libamvenc"
+DEPENDS += "libge2d"
+
+RDEPENDS_${PN} = " libge2d"
+
+inherit autotools pkgconfig
+
+SRCREV ?="${AUTOREV}"
+
+do_configure[noexec] = "1"
+
+EXTRA_OEMAKE = "TARGET_DIR=${D} STAGING_DIR=${D} DESTDIR=${D} PKG_CONFIG=${STAGING_BINDIR_NATIVE}/pkg-config"
+
+do_compile(){
+    oe_runmake -C ${S} ${EXTRA_OEMAKE} all
+}
+
+do_install() {
+    install -d -m 0755 ${D}${libdir}
+    install -d -m 0755 ${D}${libdir}/gstreamer-1.0/
+    install -D -m 0755 ${S}/libgstamlvenc.so ${D}${libdir}/gstreamer-1.0/
+}
+
+FILES_${PN} = " ${libdir}/* "
+FILES_${PN}-dev = " ${includedir}/* \
+                    /usr/lib/pkgconfig/* \
+    "
-- 
2.25.1

